File: examples/rust/ec-unit/ECUnit.runar.rs
use runar::prelude::*;

/// ECUnit -- Unit-style exercises for the secp256k1 EC built-ins.
#[runar::contract]
pub struct ECUnit {
    #[readonly]
    pub pub_key: ByteString,
}

impl ECUnit {
    /// Exercise ecMulGen, ecOnCurve, ecNegate, ecMul, ecAdd, ecPointX,
    /// ecPointY, ecMakePoint, and ecEncodeCompressed.
    pub fn test_ops(&self) {
        let g = ec_mul_gen(1);
        assert!(ec_on_curve(&g));
        let neg = ec_negate(&g);
        assert!(ec_on_curve(&neg));
        let doubled = ec_mul(&g, 2);
        assert!(ec_on_curve(&doubled));
        let sum = ec_add(&g, &g);
        assert!(ec_on_curve(&sum));
        // P + (-P) is the point at infinity: ec_add returns the all-zero blob
        // from a SUCCESSFUL script, and ec_on_curve must reject it.
        let inf_pt = ec_add(&g, &neg);
        assert!(!ec_on_curve(&inf_pt));
        let x = ec_point_x(&g);
        let y = ec_point_y(&g);
        let rebuilt = ec_make_point(x, y);
        assert!(ec_on_curve(&rebuilt));
        let compressed = ec_encode_compressed(&g);
        assert!(len(&compressed) == 33);
        assert!(true);
    }
}

File: examples/java/src/main/java/runar/examples/ec-unit/ECUnit.runar.java
package runar.examples.ecunit;

import java.math.BigInteger;

import runar.lang.SmartContract;
import runar.lang.annotations.Public;
import runar.lang.annotations.Readonly;
import runar.lang.runtime.MockCrypto.Point;
import runar.lang.types.ByteString;

import static runar.lang.Builtins.assertThat;
import static runar.lang.Builtins.ecAdd;
import static runar.lang.Builtins.ecEncodeCompressed;
import static runar.lang.Builtins.ecMakePoint;
import static runar.lang.Builtins.ecMul;
import static runar.lang.Builtins.ecMulGen;
import static runar.lang.Builtins.ecNegate;
import static runar.lang.Builtins.ecOnCurve;
import static runar.lang.Builtins.ecPointX;
import static runar.lang.Builtins.ecPointY;
import static runar.lang.Builtins.len;

/**
 * ECUnit -- unit-style exercises for the secp256k1 EC built-ins.
 *
 * <p>Stores a {@code pubKey} byte string but the actual contract logic
 * builds points off the generator and checks invariants of every EC
 * primitive in a single method. Useful as a smoke test for Bitcoin
 * Script EC opcodes.
 */
class ECUnit extends SmartContract {

    @Readonly ByteString pubKey;

    ECUnit(ByteString pubKey) {
        super(pubKey);
        this.pubKey = pubKey;
    }

    /**
     * Exercise {@code ecMulGen}, {@code ecOnCurve}, {@code ecNegate},
     * {@code ecMul}, {@code ecAdd}, {@code ecPointX}, {@code ecPointY},
     * {@code ecMakePoint}, and {@code ecEncodeCompressed}.
     */
    @Public
    void testOps() {
        Point g = ecMulGen(BigInteger.ONE);
        assertThat(ecOnCurve(g));
        Point neg = ecNegate(g);
        assertThat(ecOnCurve(neg));
        Point doubled = ecMul(g, BigInteger.TWO);
        assertThat(ecOnCurve(doubled));
        Point sum = ecAdd(g, g);
        assertThat(ecOnCurve(sum));
        BigInteger x = ecPointX(g);
        BigInteger y = ecPointY(g);
        Point rebuilt = ecMakePoint(x, y);
        assertThat(ecOnCurve(rebuilt));
        ByteString compressed = ecEncodeCompressed(g);
        assertThat(len(compressed).equals(BigInteger.valueOf(33)));
        assertThat(true);
    }
}
